src/librustc_incremental/persist/util.rs

```diff
@@ -30,3 +30,23 @@ pub fn dep_graph_path<'tcx>(tcx: &ty::TyCtxt<'tcx>) -> Option<PathBuf> {
     })
 }
 
+// Like std::fs::create_dir_all, except handles concurrent calls among multiple
+// threads or processes.
+fn create_dir_racy(path: &Path) -> io::Result<()> {
+    match fs::create_dir(path) {
+        Ok(()) => return,
+        Err(ref e) if e.kind() == io::ErrorKind::AlreadyExists => return Ok(()),
+        Err(ref e) if e.kind() == io::ErrorKind::NotFound => {}
```

Comment:
Yeah that's the case where this is emulating `create_dir_all` instead of just `create_dir`, and if you create `/foo/bar/baz` but `/foo/bar` doesn't exist you'll get `EEXIST` (not found)